pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt

```diff
@@ -643,6 +680,45 @@ class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptio
     private fun buildModifiers(modifiers: List<Modifier>) =
         ASTModifierNode(modifiers).apply { modifiers.forEach { buildAndSetParent(it, parent = this) } }
 
+    /** Generates additional nodes for the [root] node. */
+    private fun generateAdditional(root: ApexNode<*>) {
+        // Generate fields
+        findDescendants(root, nodeType = ASTFieldDeclarationStatements::class).forEach { node ->
+            generateFields(node)
+        }
+
+        findDescendants(root, nodeType = ASTProperty::class).forEach { node -> generateFields(node) }
+    }
+
+    /** Returns all descendants of [root] of type [nodeType], including [root]. */
```

Comment:
Would this be a useful utility on `Node`?

Reply:
This function operates on `ApexNode`s. Could you elaborate?

Reply:
n/m. I misread this.

w.r.t. the naming: since this also seeks to `findDescendantsOfType` like the method it wraps, I'm inclined to think it should share that name. But because it behaves differently by including the the root, that distinction seems like it should be capture in the name. any thoughts?

Reply:
Maybe this function can be removed and its uses replaced with the regular `findDescendantsOfType`?

I originally kept the "include-self" behavior for nodes like `ApexRootNode`, where it's more likely that this would be useful (e.g. when searching for `nodeType = ASTUserClass`, the root node itself could be one). This was used to generate additional methods for all `ApexRootNode`s, but currently is unimplemented because it seems to impact no rules. With the  current additional generation behaviors, it's impossible for the root node to be included.